### PR TITLE
enhance(logs): decode ANSI in log output

### PR DESCRIPTION
**Pull request by davidvader**
implemented ANSI code parser to enhance log output for tools such as Cypress
Screen Shot 2020-06-22 at 6 55 51 PM

### CODE REVIEW

**Comment by davidvader** [comment]
a lot of cypress code is abstracted out to #178, will open pull request from Draft when #178 is reviewed & merged :)

File: src/elm/Logs.elm [hunk 1]
```diff
@@ -111,25 +113,64 @@ viewLogs stepNumber logFocus log clickAction shiftDown =
...
 viewLines stepNumber logFocus log clickAction shiftDown =
-    div [ class "lines" ] <|
-        List.indexedMap
-            (\idx -> \line -> viewLine stepNumber line logFocus (idx + 1) clickAction shiftDown)
-        <|
-            decodeLogLine log
-
+    let
+        output =
+            decodeLog log
```
Comment (davidvader) [inline src/elm/Logs.elm:118]: decoding the log separate

File: src/elm/Logs.elm [hunk 2]
```diff
@@ -111,25 +113,55 @@ viewLogs stepNumber logFocus log clickAction shiftDown =
...
 viewLines stepNumber logFocus log clickAction shiftDown =
-    div [ class "lines" ] <|
-        List.indexedMap
-            (\idx -> \line -> viewLine stepNumber line logFocus (idx + 1) clickAction shiftDown)
-        <|
-            decodeLogLine log
-
-
-{-| viewLine : takes step number, line focus information, and click action and renders a log line
+    Html.table [ class "log-table" ] <|
```
Comment (davidvader) [inline src/elm/Logs.elm:116]: converting logs to <table> :)

File: src/scss/_ansi.scss [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+  background-color: darken(hsl(0, 0%, 98%), 70%);
+  border: 1px solid darken(hsl(0, 0%, 98%), 50%);
+  border-radius: 5px;
+}
+
+.ansi-black-fg {
+  color: var(--color-coal-light);
+}
+
+.ansi-red-fg {
```
Comment (davidvader) [inline src/scss/_ansi.scss:26]: these are the juice of the PR
I tried to re-use as many of the default colors as I could for the ANSI colors

File: src/scss/_ansi.scss [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+// Use of this source code is governed by the LICENSE file in this repository.
+
+/* ansi text */
+.ansi-bold {
+  font-weight: bold;
+}
+
+.ansi-faint {
+  // cannot use sass variables in functions
+  color: darken(hsl(0, 0%, 98%), 50%);
```
Comment (davidvader) [inline src/scss/_ansi.scss:12]: I couldn't use a func substitution variable here... :/

File: src/scss/_ansi.scss [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+.ansi-blink {
+  animation: 1s blinker step-end infinite;
+}
+
+.monospace-bold {
+  font-weight: bold;
+  font-family: monospace;
+}
+
+@keyframes blinker {
```
Comment (davidvader) [inline src/scss/_ansi.scss:183]: could move this to _animations.css, but its small and only related to ansi

File: src/scss/_buttons.scss [hunk 6]
```diff
@@ -227,3 +227,15 @@
...
+button.-ptr:hover,
+button.-ptr:focus {
+  cursor: pointer;
+}
+
+// button underline style
+button.-link-ul:hover,
+button.-link-ul:focus {
+  text-decoration: underline;
+}
```
Comment (davidvader) [inline src/scss/_buttons.scss:241]: making the clickability of the log line number button more obvious 👍

File: src/scss/_main.scss [hunk 7]
```diff
@@ -1490,3 +1485,11 @@ nav {
 .visually-hidden {
   @include visually-hidden();
 }
+
```
Comment (davidvader) [inline src/scss/_main.scss:1488]: wasnt sure where else to put these, so I put them at the bottom of `_main.scss`

File: src/elm/Logs.elm [hunk 8]
```diff
@@ -469,3 +493,53 @@ toString log =
...
             ""
+
+
+
+-- ANSI
+
+
+{-| defaultLogModel : struct to represent default model required by ANSI parser
+-}
+defaultLogModel : Ansi.Log.Model
```
Comment (davidvader) [inline src/elm/Logs.elm:504]: models required by the ansi parser package
see:
https://package.elm-lang.org/packages/vito/elm-ansi/latest/Ansi-Log

File: src/elm/Logs.elm [hunk 9]
```diff
@@ -469,3 +493,53 @@ toString log =
...
+    { row = 0
+    , column = 0
+    }
+
+
+{-| decodeAnsi : takes maybe log parses into ansi decoded log line array
+-}
+decodeAnsi : Maybe (WebData Log) -> Array.Array Ansi.Log.Line
+decodeAnsi log =
+    .lines <| Ansi.Log.update (decodeLog log) defaultLogModel
```
Comment (davidvader) [inline src/elm/Logs.elm:545]: this is where the magic happens 🤘 
takes a default empty ansi log model, parses out the encoded style, and returns a list of lines with applied style actions, ready to be rendered

see: https://package.elm-lang.org/packages/vito/elm-ansi/latest/Ansi-Log#update

File: src/elm/Logs.elm [hunk 10]
```diff
@@ -111,25 +113,55 @@ viewLogs stepNumber logFocus log clickAction shiftDown =
...
+            , Util.testAttribute <| String.join "-" [ "log", "line", stepNumber, String.fromInt lineNo ]
+            , logFocusStyles logFocus lineNo
             ]
-            [ lineFocusButton stepNumber logFocus lineNumber clickAction shiftDown
-            , code [] [ text line ]
+            [ Html.td []
+                [ lineFocusButton stepNumber logFocus lineNo setLogFocus shiftDown ]
+            , Html.td [ class "-word-break-all", class "-overflow-auto" ]
+                [ code [ Util.testAttribute <| String.join "-" [ "log", "data", stepNumber, String.fromInt lineNo ] ]
+                    [ Ansi.Log.viewLine line ]
```
Comment (davidvader) [inline src/elm/Logs.elm:163]: this is where you take a decoded ansi log line and render it with the appropriate style

File: src/elm/Logs.elm (hunk 9 above)
Comment (davidvader) [inline src/elm/Logs.elm:545]: this is where the magic happens 🤘
takes a default empty ansi log model, parses out the encoded style, and returns a list of lines with applied style actions, ready to be rendered

see: https://package.elm-lang.org/packages/vito/elm-ansi/latest/Ansi-Log#update

File: src/scss/_ansi.scss [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+/* ansi text */
+.ansi-bold {
+  font-weight: bold;
+}
+
+.ansi-faint {
+  // cannot use sass variables in functions
+  color: darken(hsl(0, 0%, 98%), 50%);
+}
+
```
Comment (davidvader) [inline src/scss/_ansi.scss:14]: I couldn't use a func substitution variable here... :/

File: src/scss/_ansi.scss [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+@keyframes blinker {
+  from,
+  to {
+    color: transparent;
+  }
+
+  50% {
+    color: inherit;
+  }
+}
```
Comment (davidvader) [inline src/scss/_ansi.scss:192]: could move this to _animations.scss, but its small and only related to ansi

File: src/scss/_buttons.scss (hunk 6 above)
Comment (davidvader) [inline src/scss/_buttons.scss:241]: making the clickability of the log line number button more obvious 👍

File: src/scss/_main.scss [hunk 13]
```diff
@@ -1490,3 +1485,11 @@ nav {
...
+.-word-break-all {
+  word-break: break-all;
+}
+
+.-overflow-auto {
+  overflow: auto;
+}
```
Comment (davidvader) [inline src/scss/_main.scss:1495]: wasnt sure where else to put these, so I put them at the bottom of _main.scss

File: src/elm/Logs.elm [hunk 14]
```diff
@@ -142,9 +174,11 @@ lineFocusButton stepNumber logFocus lineNumber clickAction shiftDown =
         [ Util.onClickPreventDefault <|
             clickAction <|
                 logRangeId stepNumber lineNumber logFocus shiftDown
-        , Util.testAttribute <| "log-line-num-" ++ String.fromInt lineNumber
+        , Util.testAttribute <| String.join "-" [ "log", "line", "num", stepNumber, String.fromInt lineNumber ]
         , id <| stepAndLineToFocusId stepNumber lineNumber
         , class "line-number"
+        , class "-link-ul"
+        , class "-ptr"
```
Comment (davidvader) [inline src/elm/Logs.elm:181]: making clickability of the log line numbers a bit more obvious 👍

File: src/scss/_ansi.scss [hunk 15]
```diff
@@ -0,0 +1,192 @@
...
+  background-color: var(--color-offwhite);
+}
+
+.ansi-blink {
+  animation: 1s blinker step-end infinite;
+}
+
+.monospace-bold {
+  font-weight: bold;
+  font-family: monospace;
```
Comment (wass3r) [inline src/scss/_ansi.scss:180]: We should probably reference https://github.com/go-vela/ui/blob/master/src/scss/_variables.scss#L50 to stay consistent.

File: src/elm/Logs.elm (hunk 14 above)
Comment (wass3r) [reply src/elm/Logs.elm:170]: i feel like an existing `.button` class w/ modifier might work here. would `.button.-link` get us there? it has both underline and pointer.

File: src/scss/_ansi.scss (hunk 11 above)
Comment (wass3r) [reply src/scss/_ansi.scss:14]: i would propose that we should probably still define these in `_variables` so we have all the colors defined in one place.

File: src/elm/Logs.elm (hunk 14 above)
Comment (davidvader) [reply src/elm/Logs.elm]: yeah it added some inadvertent style but I still like it :) 
Screen Shot 2020-07-03 at 12 40 17 PM

Comment (davidvader) [reply src/elm/Logs.elm]: updated 👍

File: src/scss/_ansi.scss (hunk 15 above)
Comment (davidvader) [reply src/scss/_ansi.scss]: agreed i missed that one 👍 updated the colors as well

**Comment by davidvader** [comment]
@wass3r updated 👍

**Pull request merged**